File: src/main/java/io/pivotal/kotlindemo/java/clients/ClientsController.java
package io.pivotal.kotlindemo.java.clients;

import io.pivotal.kotlindemo.java.accounts.Account;
import io.pivotal.kotlindemo.java.accounts.AccountRepository;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.*;

import java.util.List;

@RestController
@RequestMapping("/api/clients")
public class ClientsController {

    private final GetClientOverviews getClientOverviews;
    private final ClientRepository clientRepository;
    private final AccountRepository accountRepository;

    public ClientsController(
        GetClientOverviews getClientOverviews,
        ClientRepository clientRepository,
        AccountRepository accountRepository
    ) {
        this.getClientOverviews = getClientOverviews;
        this.clientRepository = clientRepository;
        this.accountRepository = accountRepository;
    }

    @GetMapping
    public ResponseEntity<List<ClientOverview>> getAll() {
        return ResponseEntity.ok(getClientOverviews.execute());
    }

    @GetMapping("/{id}")
    public ResponseEntity<Client> getById(@PathVariable Long id) {
        try {
            return ResponseEntity.ok(clientRepository.findById(id));
        } catch (NoClientFoundException e) {
            return ResponseEntity.notFound().build();
        }
    }

    @GetMapping("/{clientId}/accounts")
    public ResponseEntity<List<Account>> getAccountsForClientId(@PathVariable Long clientId) {
        return ResponseEntity.ok(accountRepository.findByClientId(clientId));
    }

//    @GetMapping
//    public ResponseEntity<List<Client>> getByLastName(@RequestParam String lastName) {
//        return ResponseEntity.ok(clientRepository.findByLastName(lastName));
//    }
//
//    @GetMapping
//    public ResponseEntity<Client> getByEmail(@RequestParam String email) {
//        Optional<Client> clientOptional = clientRepository.findByEmail(email);
//        if (clientOptional.isPresent()) {
//            return ResponseEntity.ok(clientOptional.get());
//        } else {
//            return ResponseEntity.notFound().build();
//        }
//    }

    @PostMapping
    public ResponseEntity<Client> create(@RequestBody CreateClientRequest createClientRequest) {
        return ResponseEntity.ok(clientRepository.create(createClientRequest));
    }

}

File: src/main/kotlin/io/pivotal/kotlindemo/kotlin/clients/ClientsController.kt
package io.pivotal.kotlindemo.kotlin.clients

import io.pivotal.kotlindemo.kotlin.accounts.Account
import io.pivotal.kotlindemo.kotlin.accounts.AccountRepository
import org.springframework.http.ResponseEntity
import org.springframework.web.bind.annotation.*

@RestController
@RequestMapping("/api/clients")
class ClientsController(
    private val getClientOverviews: GetClientOverviews,
    private val clientRepository: ClientRepository,
    private val accountRepository: AccountRepository
) {

    @GetMapping
    fun getAll() = ResponseEntity.ok(getClientOverviews.execute())

    @GetMapping("/{id}")
    fun getById(@PathVariable id: Long): ResponseEntity<Client> {
        try {
            return ResponseEntity.ok(clientRepository.findById(id))
        } catch (e: NoClientFoundException) {
            return ResponseEntity.notFound().build()
        }

    }

    @GetMapping("/{clientId}/accounts")
    fun getAccountsForClientId(@PathVariable clientId: Long): ResponseEntity<List<Account>> {
        return ResponseEntity.ok(accountRepository.findByClientId(clientId))
    }

//    @GetMapping
//    fun getByLastName(@RequestParam lastName: String) =
//        ResponseEntity.ok(clientRepository.findByLastName(lastName))
//
//    @GetMapping
//    fun getByEmail(@RequestParam email: String): ResponseEntity<Client> {
//        val client = clientRepository.findByEmail(email)
//        client ?: return ResponseEntity.notFound().build()
//        return ResponseEntity.ok(client)
//    }

    @PostMapping
    fun create(@RequestBody createClientRequest: CreateClientRequest) =
        ResponseEntity.ok(clientRepository.create(createClientRequest))

}
